### PR TITLE
Support rsp in MTP

### LINKED ISSUE #3730: Support `@response.rsp` files
Command line can contains long arguments (i.e. `--filter`).
Allow to pass a response file will help these scenario.

AB#2302861

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #3730

### CODE REVIEW

File: src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+
+using Microsoft.Testing.Platform.Resources;
+
+// Most of the core logic is from https://github.com/dotnet/command-line-api/blob/feb61c7f328a2401d74f4317b39d02126cfdfe24/src/System.CommandLine/Parsing/CliParser.cs#L49
```
Comment (MarcoRossignoli) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs]: you meant here https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Parsing/StringExtensions.cs#L349?

Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs]: It's a mix actually 😄

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/CommandLine/CommandLineTests.cs [hunk 2]
```diff
@@ -10,65 +10,125 @@ namespace Microsoft.Testing.Platform.UnitTests;
...
+            --option4 d
+            """)], new CommandLineParseResultWrapper(null, new List<OptionRecord>
+        {
+            new("option1", ["a"]),
+            new("option2", ["b"]),
+            new("option3", ["c"]),
+            new("option4", ["d"]),
+            new("option5", ["e"]),
+            new("option6", ["f"]),
+        }.ToArray(), []));
```
Comment (MarcoRossignoli) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/CommandLine/CommandLineTests.cs:131]: pls add a pair of more sample with 
1) rsp file at the start 
2) rsp file t the end
3) only rsp file
4) 2 rsp file only

**Review by MarcoRossignoli — APPROVED** [review]
LGTM :shipit:

File: src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+using System.Globalization;
+
+using Microsoft.Testing.Platform.Resources;
+
+// Most of the core logic is from:
+// - https://github.com/dotnet/command-line-api/blob/feb61c7f328a2401d74f4317b39d02126cfdfe24/src/System.CommandLine/Parsing/CliParser.cs#L49
+// - https://github.com/dotnet/command-line-api/blob/feb61c7f328a2401d74f4317b39d02126cfdfe24/src/System.CommandLine/Parsing/StringExtensions.cs#L349
+internal static class ResponseFileHelper
+{
+    internal static bool TryReadResponseFile(string rspFilePath, List<string> errors, [NotNullWhen(true)] out string[]? newArguments)
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs]: Nit: let's use higher abstraction for errors.

```suggestion
    internal static bool TryReadResponseFile(string rspFilePath, ICollection<string> errors, [NotNullWhen(true)] out string[]? newArguments)
```

File: src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        }
+        catch (IOException e)
+        {
+            errors.Add(string.Format(CultureInfo.InvariantCulture, PlatformResources.CommandLineParserFailedToReadResponseFile, rspFilePath, e.Message));
+        }
+
+        newArguments = null;
+        return false;
+
+        static IEnumerable<string> ExpandResponseFile(string filePath)
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs:34]: Stylistic nit:

```suggestion
        // Local functions
        static IEnumerable<string> ExpandResponseFile(string filePath)
```

File: src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        {
+            string[] lines = File.ReadAllLines(filePath);
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs:36]: Could it be worth reading async (cc @MarcoRossignoli)?

File: src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+            newArguments = ExpandResponseFile(rspFilePath).ToArray();
+            return true;
+        }
+        catch (FileNotFoundException)
+        {
+            errors.Add(string.Format(CultureInfo.InvariantCulture, PlatformResources.CommandLineParserResponseFileNotFound, rspFilePath));
+        }
+        catch (IOException e)
+        {
+            errors.Add(string.Format(CultureInfo.InvariantCulture, PlatformResources.CommandLineParserFailedToReadResponseFile, rspFilePath, e.Message));
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs]: It might be worth adding the tostring of the exception.

Comment (MarcoRossignoli) [reply src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs:27]: We could also as follow-up if we don't want to change all the callers/tests right now.

File: src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs (hunk 5 above)
Comment (MarcoRossignoli) [reply src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs:36]: We could also as follow-up if we don't want to change all the callers/tests right now.

Comment (Youssef1313) [reply src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs:36]: It's possible to make it async but I don't see much benefit. We will follow-up if needed

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/CommandLine/CommandLineTests.cs (hunk 2 above)
Comment (Youssef1313) [reply test/UnitTests/Microsoft.Testing.Platform.UnitTests/CommandLine/CommandLineTests.cs:131]: Done

**Comment by Youssef1313** [comment]
Thanks for the reviews @MarcoRossignoli @Evangelink. I addressed your feedback

File: src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs (hunk 5 above)
Comment (Evangelink) [reply src/Platform/Microsoft.Testing.Platform/CommandLine/ResponseFileHelper.cs:36]: @Youssef1313 the idea is for response file that would contain hundred of thousands of lines. As you both said, fine for me as follow-up.

**Review by Evangelink — APPROVED** [review]

**Pull request merged**